README.md
# @valbo/ajv-strict-formats

Stricter versions of JSON Schema formats for Ajv.

![npm (scoped)](https://img.shields.io/npm/v/@valbo/ajv-strict-formats)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
![Build Status](https://img.shields.io/github/workflow/status/valverdealbo/ajv-strict-formats/CI)
[![Coverage Status](https://coveralls.io/repos/github/valverdealbo/ajv-strict-formats/badge.svg?branch=main)](https://coveralls.io/github/valverdealbo/ajv-strict-formats?branch=main)
[![Known Vulnerabilities](https://snyk.io/test/github/valverdealbo/ajv-strict-formats/badge.svg?targetFile=package.json)](https://snyk.io/test/github/valverdealbo/ajv-strict-formats?targetFile=package.json)

## Install

```bash
npm install @valbo/ajv-strict-formats
```

## Usage

Adds **time**, **date-time** and **uuid** formats to [Ajv](https://github.com/ajv-validator/ajv) that are stricter than the ones from the [ajv-formats](https://github.com/ajv-validator/ajv-formats) package:

```typescript
import Ajv from 'ajv';
import addFormats from '@valbo/ajv-strict-formats';

const ajv = new Ajv();
addFormats(ajv);
```

Or add only some formats:

```typescript
import Ajv from 'ajv';
import addFormats from '@valbo/ajv-strict-formats';

const ajv = new Ajv();
addFormats(ajv, ['date-time', 'uuid']);
```

### time format

A stricter format that only accepts **HH:mm:ss** with no fractions of seconds or timezone.

### date-time format

A stricter format that only accepts **T** as a separator and **Z** as the timezone: **2021-02-13T22:04:00Z**

Fractions of seconds are allowed: **2021-02-13T22:04:00.000Z**

### uuid format

A stricter format that only allows lowercase characters.


CHANGELOG.md
## [1.0.5](https://github.com/valverdealbo/ajv-strict-formats/compare/v1.0.4...v1.0.5) (2021-04-17)


### Bug Fixes

* change ajv dependency to peer dependency ([9492bbf](https://github.com/valverdealbo/ajv-strict-formats/commit/9492bbf102d0be3aac69b49d418b3a2ea89385b2))

## [1.0.4](https://github.com/valverdealbo/ajv-strict-formats/compare/v1.0.3...v1.0.4) (2021-04-17)


### Bug Fixes

* update dependencies, update tools configurations ([d53fff0](https://github.com/valverdealbo/ajv-strict-formats/commit/d53fff07774487ea0e001504c19b1f7bcc051ec6))

## [1.0.3](https://github.com/valverdealbo/ajv-strict-formats/compare/v1.0.2...v1.0.3) (2021-03-14)


### Bug Fixes

* upgrade ajv from 7.1.0 to 7.1.1 ([e9208d5](https://github.com/valverdealbo/ajv-strict-formats/commit/e9208d58c9e2dd890f133c7d7e1cceb49423b3d4))

## [1.0.2](https://github.com/valverdealbo/ajv-strict-formats/compare/v1.0.1...v1.0.2) (2021-02-19)


### Bug Fixes

* update license to MIT ([b8f142f](https://github.com/valverdealbo/ajv-strict-formats/commit/b8f142ff0d12b73e44a3a7d6aba91442282888df))

## [1.0.1](https://github.com/valverdealbo/ajv-strict-formats/compare/v1.0.0...v1.0.1) (2021-02-17)


### Bug Fixes

* remove date format ([f6e945f](https://github.com/valverdealbo/ajv-strict-formats/commit/f6e945f20d89140d6ba6e0187b7c098784f409ae))


src/index.test.ts
import Ajv from 'ajv';
import { time, datetime, uuid } from './formats';
import addFormats from '.';

describe('addFormats()', () => {
  test('should add all formats', () => {
    const ajv = new Ajv();
    addFormats(ajv);
    expect(ajv.formats.time).toBeDefined();
    expect(ajv.formats['date-time']).toBeDefined();
    expect(ajv.formats.uuid).toBeDefined();
  });

  test('should add only the provided formats', () => {
    const ajv = new Ajv();
    addFormats(ajv, ['time']);
    expect(ajv.formats.time).toBeDefined();
    expect(ajv.formats['date-time']).toBeUndefined();
    expect(ajv.formats.uuid).toBeUndefined();
  });

  test('should return the right validation function', () => {
    expect(addFormats.get('time')).toBe(time);
    expect(addFormats.get('date-time')).toBe(datetime);
    expect(addFormats.get('uuid')).toBe(uuid);
  });
});


test/index.test.ts
import Ajv from 'ajv';
import { time, datetime, uuid } from '../src/formats';
import addFormats from '../src';

describe('addFormats()', () => {
  test('should add all formats', () => {
    const ajv = new Ajv();
    addFormats(ajv);
    expect(ajv.formats.time).toBeDefined();
    expect(ajv.formats['date-time']).toBeDefined();
    expect(ajv.formats.uuid).toBeDefined();
  });

  test('should add only the provided formats', () => {
    const ajv = new Ajv();
    addFormats(ajv, ['time']);
    expect(ajv.formats.time).toBeDefined();
    expect(ajv.formats['date-time']).toBeUndefined();
    expect(ajv.formats.uuid).toBeUndefined();
  });

  test('should return the right validation function', () => {
    expect(addFormats.get('time')).toBe(time);
    expect(addFormats.get('date-time')).toBe(datetime);
    expect(addFormats.get('uuid')).toBe(uuid);
  });
});


src/formats.test.ts
import Ajv from 'ajv';
import { time, datetime, uuid, formats } from './formats';

describe('time()', () => {
  const ajv = new Ajv();
  ajv.addFormat('time', time);
  const schema = { type: 'string', format: 'time' };

  test('should return invalid format if the time does not match', () => {
    ajv.validate(schema, '9:00:00');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format when there is a timezone', () => {
    ajv.validate(schema, '09:00:00Z');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format when there are seconds fraction', () => {
    ajv.validate(schema, '09:00:00.000');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return valid format', () => {
    ajv.validate(schema, '09:00:00');
    expect(ajv.errors).toBeNull();
  });

  test('should return valid format for a leap second ', () => {
    ajv.validate(schema, '23:59:60');
    expect(ajv.errors).toBeNull();
  });
});

describe('datetime()', () => {
  const ajv = new Ajv();
  ajv.addFormat('date-time', datetime);
  const schema = { type: 'string', format: 'date-time' };

  test('should return invalid format if the date does not match', () => {
    ajv.validate(schema, '200-02-29T09:00:00Z');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format if the time does not match', () => {
    ajv.validate(schema, '2000-02-29T9:00:00Z');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format when the separator is not a "T"', () => {
    ajv.validate(schema, '2019-01-01 09:00:00Z');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format when there is no timezone', () => {
    ajv.validate(schema, '2019-01-01T09:00:00');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return valid format when there are no seconds fraction', () => {
    ajv.validate(schema, '2019-01-01T09:00:00Z');
    expect(ajv.errors).toBeNull();
  });

  test('should return valid format when there are seconds fraction', () => {
    ajv.validate(schema, '2019-01-01T09:00:00.000Z');
    expect(ajv.errors).toBeNull();
  });

  test('should return valid format for a leap year', () => {
    ajv.validate(schema, '2016-02-29T09:00:00Z');
    expect(ajv.errors).toBeNull();
  });

  test('should return valid format for a century leap year', () => {
    ajv.validate(schema, '2000-02-29T09:00:00Z');
    expect(ajv.errors).toBeNull();
  });

  test('should return valid format for a leap second ', () => {
    ajv.validate(schema, '2000-02-29T23:59:60Z');
    expect(ajv.errors).toBeNull();
  });
});

describe('uuid()', () => {
  const ajv = new Ajv();
  ajv.addFormat('uuid', uuid);
  const schema = { type: 'string', format: 'uuid' };

  test('should return invalid format when the value is not an uuid', () => {
    ajv.validate(schema, '87b494c9-68e3-4281-95b8-18ddde05c71b6');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return invalid format when there are uppercase', () => {
    ajv.validate(schema, '87B494C9-68E3-4281-95B8-8DDDE05C71B6');
    expect(ajv.errors).not.toBeNull();
  });

  test('should return valid format when there are only lowercase', () => {
    ajv.validate(schema, '87b494c9-68e3-4281-95b8-8ddde05c71b6');
    expect(ajv.errors).toBeNull();
  });
});

describe('formats', () => {
  test('each key should the right validation function', () => {
    expect(formats.time).toBe(time);
    expect(formats['date-time']).toBe(datetime);
    expect(formats.uuid).toBe(uuid);
  });
});


src/formats.ts
function isLeapYear(year: number): boolean {
  return year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
}

const DATE = /^(\d\d\d\d)-(\d\d)-(\d\d)$/;
const DAYS = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

function date(str: string): boolean {
  const matches: string[] | null = DATE.exec(str);
  if (!matches) return false;
  const year: number = +matches[1];
  const month: number = +matches[2];
  const day: number = +matches[3];
  return month >= 1 && month <= 12 && day >= 1 && day <= (month === 2 && isLeapYear(year) ? 29 : DAYS[month]);
}

const TIME = /^(\d\d):(\d\d):(\d\d)$/;
const FULL_TIME = /^(\d\d):(\d\d):(\d\d)(\.\d+)?Z$/;

export function time(str: string, regexp: RegExp = TIME): boolean {
  const matches: string[] | null = regexp.exec(str);
  if (!matches) return false;
  const hour = +matches[1];
  const minute = +matches[2];
  const second = +matches[3];
  return (hour <= 23 && minute <= 59 && second <= 59) || (hour === 23 && minute === 59 && second === 60);
}

const DATE_TIME_SEPARATOR = /T/;

export function datetime(str: string): boolean {
  const dateTime: string[] = str.split(DATE_TIME_SEPARATOR);
  return dateTime.length === 2 && date(dateTime[0]) && time(dateTime[1], FULL_TIME);
}

const UUID = /^[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}$/;

export function uuid(str: string): boolean {
  return UUID.exec(str) !== null;
}

export type FormatName = 'time' | 'date-time' | 'uuid';

export type Formats = { [K in FormatName]: (str: string) => boolean };

export const formats: Formats = {
  time,
  'date-time': datetime,
  uuid,
};

export const formatNames = Object.keys(formats) as FormatName[];
